converters/toZigbee.js

```diff
@@ -841,234 +842,98 @@ const converters = {
-                value.brightness = utils.mapNumberRange(hsv.v, 0, 100, 0, 254);
-                newState.brightness = value.brightness;
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturationAndBrightness';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
-                }
-            } else if (value.hasOwnProperty('hsb')) {
-                let hsv = value.hsb.split(',').map((i) => parseInt(i));
-                newState.color = {h: hsv[0], s: hsv[1], v: hsv[2]};
-                hsv = utils.gammaCorrectHSV(utils.correctHue(hsv[0], meta), hsv[1], hsv[2]);
-                value.saturation = utils.mapNumberRange(hsv.s, 0, 100, 0, 254);
-                value.brightness = utils.mapNumberRange(hsv.v, 0, 100, 0, 254);
-                newState.brightness = value.brightness;
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturationAndBrightness';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
-                }
-            } else if (value.hasOwnProperty('h') && value.hasOwnProperty('s') && value.hasOwnProperty('v')) {
-                newState.color = {h: value.h, s: value.s, v: value.v};
-                const hsv = utils.gammaCorrectHSV(utils.correctHue(value.h, meta), value.s, value.v);
-                value.saturation = utils.mapNumberRange(hsv.s, 0, 100, 0, 254);
-                value.brightness = utils.mapNumberRange(hsv.v, 0, 100, 0, 254);
-                newState.brightness = value.brightness;
-                if (utils.getMetaValue(entity, meta.mapped, 'enhancedHue', 'allEqual', true)) {
-                    value.hue = utils.mapNumberRange(hsv.h % 360, 0, 360, 0, 65535);
-                    command = 'enhancedMoveToHueAndSaturationAndBrightness';
-                } else {
-                    value.hue = utils.mapNumberRange(hsv.h, 0, 360, 0, 254);
-                    command = 'moveToHueAndSaturationAndBrightness';
+            if (newColor.rgb !== null || newColor.xy !== null) {
+                // Convert RGB to XY color mode because at the moment all known devices support this mode.
+                newState.color_mode = constants.colorMode[1];
+                let xy;
+
+                if (newColor.rgb !== null) {
+                    const rgb = newColor.rgb;
+                    // XXX: store raw RGB, raw XY or corrected XY?
```

Comment:
The current behaviour when setting via RGB is that the corrected XY is returned in the state. We only send the supported color modes in the ZCL: hue/saturation, x/y and color temperature (CC: @sjorge)

Reply:
That is what I initially assumed yes, but this does not look to be true.

```
mosquitto_pub -t zigbee2mqtt/0xXXXXXXX/set/color -m '{"h": 120, "s": 100}'
```

Will currently add save/publish a color state entry with `h` and `s` set. While I expected a color state with `hue` and `saturation` set.

Base from the HA issue noted earlier, I think this was done because HA expect h and s to be present when using a hs color mode. It does not check for `hue` and `saturation`... which also explains why physical remotes that set hue and saturation, when reported or read back don't update HA.

Reply:
ok, done